aten/src/ATen/native/cuda/LinearAlgebra.cu

```diff
@@ -178,4 +164,207 @@ Tensor mm_cuda(const Tensor& self, const Tensor& mat2) {
+              out.device() == vec1.device() &&
+              out.device() == vec2.device(),
+              "Expected all tensors to be on the same device. Found: ",
+              out.device(), ", ", self_.device(), ", ",
+              vec1.device(), " and ", vec2.device());
+  TORCH_CHECK(self_.dim() == 2,
+              "2D tensor expected, got ", self_.dim(), "D tensor for input");
+  TORCH_CHECK(self_.size(0) == vec1.size(0) && self_.size(1) == vec2.size(0),
+              "size mismatch",
+              ", input: ", self_.sizes(),
+              ", v1: ", vec1.sizes(),
+              ", v2: ", vec2.sizes());
+  AT_DISPATCH_FLOATING_TYPES_AND2(kBFloat16, kHalf, self_.scalar_type(), "addr_out_cuda", [&] {
+      addr_impl_cuda<scalar_t>(out, self_, vec1, vec2,
+                               alpha.to<scalar_t>(), beta.to<scalar_t>());
+  });
+  return out;
+}
+
+Tensor& addr__cuda(Tensor& self,
+                   const Tensor& vec1, const Tensor& vec2,
+                   Scalar beta, Scalar alpha) {
+  addr_out_cuda(self, self, vec1, vec2, beta, alpha);
+  return self;
+}
+
+Tensor addr_cuda(const Tensor& self,
+                  const Tensor& vec1, const Tensor& vec2,
+                  Scalar beta, Scalar alpha) {
+  Tensor out = at::empty({0}, self.options());
+  addr_out_cuda(out, self, vec1, vec2, beta, alpha);
+  return out;
+}
+
+Tensor& addbmm_out_cuda(Tensor& out, const Tensor& self,
+                        const Tensor& batch1, const Tensor& batch2,
+                        Scalar beta, Scalar alpha) {
+  TORCH_CHECK(batch1.dim() == 3 && batch2.dim() == 3,
+              "Batch tensors should be 3D, got dimensions ", batch1.dim(),
+              " and ", batch2.dim());
+
+  Tensor self_;
+  if (&out != &self) {
+    std::tie(self_) = expand_size(self, {batch1.size(1), batch2.size(2)}, "addbmm");
+  } else {
+    self_ = self;
+  }
+
+  TORCH_CHECK(out.device() == self_.device() &&
+              out.device() == batch1.device() &&
+              out.device() == batch2.device(),
+              "Expected all tensors to be on the same device. Found: ",
+              out.device(), ", ", self_.device(), ", ",
+              batch1.device(), " and ", batch2.device());
+  TORCH_CHECK(self_.dim() == 2,
+              "2D tensor expected, got ", self_.dim(), "D tensor for input");
+  int64_t batchnum = batch1.size(0);
+  int64_t m1d1 = batch1.size(1);
+  int64_t innerdim = batch1.size(2);
+  int64_t m2d2 = batch2.size(2);
+  TORCH_CHECK(batchnum == batch2.size(0),
+              "equal number of batches expected");
+  TORCH_CHECK(m1d1 == self_.size(0),
+              "first dimension of batch1  must match first dimension of input");
+  TORCH_CHECK(m2d2 == self_.size(1),
+              "second dimension of batch2 must match second dimension of input");
+  TORCH_CHECK(innerdim == batch2.size(1),
+              "second dimension of batch1 must match first dimension of batch2");
+
+  if (&out != &self) {
```

Comment:
You might want to make this slightly more robust so it handles the case where `out` and `self` are different tensors but they have the same view of the same data, with something like this: `(out.data_ptr() != self.data_ptr()) || (out.strides() != self.strides())`.

Reply:
Are you sure that's the desired semantics of broadcasting? In https://pytorch.org/docs/stable/notes/broadcasting.html I see only that in-place broadcasting is disabled.

Reply:
Ah, you're right! (Sorry to take so long to respond, I'm not getting email notifications on this PR for some reason)